research/flamby/fed_isic2019/ditto_mkmmd/client.py

```diff
@@ -0,0 +1,203 @@
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+from torch.utils.data import DataLoader
+
+from fl4health.checkpointing.checkpointer import BestLossTorchCheckpointer
+from fl4health.checkpointing.client_module import ClientCheckpointModule
+from fl4health.clients.mkmmd_clients.ditto_mkmmd_client import DittoMkmmdClient
+from fl4health.utils.losses import LossMeterType
+from fl4health.utils.metrics import BalancedAccuracy, Metric
+from fl4health.utils.random import set_all_random_seeds
+from research.flamby.flamby_data_utils import construct_fedisic_train_val_datasets
+
+FED_ISIC2019_BASELINE_LAYERS = []
+for i in range(16):
+    FED_ISIC2019_BASELINE_LAYERS.append(f"base_model._blocks.{i}")
+FED_ISIC2019_BASELINE_LAYERS += ["base_model._dropout"]
+
+
+class FedIsic2019DittoClient(DittoMkmmdClient):
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        client_number: int,
+        learning_rate: float,
+        lam: float = 0,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        mkmmd_loss_weight: float = 10,
+        feature_l2_norm_weight: float = 1,
+        mkmmd_loss_depth: int = 1,
+        beta_global_update_interval: int = 20,
+        checkpointer: Optional[ClientCheckpointModule] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            checkpointer=checkpointer,
+            lam=lam,
+            mkmmd_loss_weight=mkmmd_loss_weight,
+            flatten_feature_extraction_layers={
+                key: True for key in FED_ISIC2019_BASELINE_LAYERS[-1 * mkmmd_loss_depth :]
+            },
+            feature_l2_norm_weight=feature_l2_norm_weight,
+            beta_global_update_interval=beta_global_update_interval,
+        )
+        self.client_number = client_number
+        self.learning_rate: float = learning_rate
+
+        assert 0 <= client_number < NUM_CLIENTS
+        log(INFO, f"Client Name: {self.client_name}, Client Number: {self.client_number}")
+
+    def get_data_loaders(self, config: Config) -> Tuple[DataLoader, DataLoader]:
+        train_dataset, validation_dataset = construct_fedisic_train_val_datasets(
+            self.client_number, str(self.data_path)
+        )
+        train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
+        val_loader = DataLoader(validation_dataset, batch_size=BATCH_SIZE, shuffle=False)
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        model: nn.Module = Baseline().to(self.device)
+        return model
+
+    def get_optimizer(self, config: Config) -> Dict[str, Optimizer]:
+        # Note that the global optimizer operates on self.global_model.parameters() and
```

Comment:
I'm sure this is copy and pasted from somewhere that I didn't complete this comment properly, but this comment ends on an and 🙂

Reply:
This is copied from original ditto which it is missing there too. I will update both.

Reply:
They are also missing in fed_heart_disease and fed_ixi examples.

Reply:
Yeah I figured this was my fault 😂. Thanks for fixing it 🙂